src/uct/ib/base/ib_device.c

```diff
@@ -340,99 +340,120 @@ uint8_t uct_ib_to_fabric_time(double time)
     }
 }
 
-void uct_ib_address_pack(uct_ib_device_t *dev, uct_ib_address_scope_t scope,
+void uct_ib_address_pack(uct_ib_device_t *dev, uct_ib_address_type_t type,
                          const union ibv_gid *gid, uint16_t lid,
                          uct_ib_address_t *ib_addr)
 {
     void *ptr = ib_addr + 1;
 
     ib_addr->flags  = 0;
 
-    /* LID */
-    /* TODO check IB link layer of the port */
-    ib_addr->flags |= UCT_IB_ADDRESS_FLAG_LID;
-    *(uint16_t*)ptr = lid;
-    ptr += sizeof(uint16_t);
-
-    if (scope >= UCT_IB_ADDRESS_SCOPE_SITE_LOCAL) {
-        ib_addr->flags |= UCT_IB_ADDRESS_FLAG_IF_ID;
-        *(uint64_t*)ptr = gid->global.interface_id;
-        ptr += sizeof(uint64_t);
-
-        if (scope >= UCT_IB_ADDRESS_SCOPE_GLOBAL) {
-            /* Global */
-            ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET64;
-            *(uint64_t*)ptr = gid->global.subnet_prefix;
-        } else {
-            /* Site-local */
-            ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET16;
-            *(uint16_t*)ptr = gid->global.subnet_prefix >> 48;
+    if (type < UCT_IB_ADDRESS_TYPE_ETH_LOCAL) {
+        /* IB */
+        /* LID */
+        ib_addr->flags |= UCT_IB_ADDRESS_FLAG_LID;
+        *(uint16_t*) ptr = lid;
+        ptr += sizeof(uint16_t);
+
+        if (type >= UCT_IB_ADDRESS_TYPE_SITE_LOCAL) {
+            ib_addr->flags |= UCT_IB_ADDRESS_FLAG_IF_ID;
+            *(uint64_t*) ptr = gid->global.interface_id;
+            ptr += sizeof(uint64_t);
+
+            if (type >= UCT_IB_ADDRESS_TYPE_GLOBAL) {
+                /* Global */
+                ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET64;
+                *(uint64_t*) ptr = gid->global.subnet_prefix;
+            } else {
+                /* Site-local */
+                ib_addr->flags |= UCT_IB_ADDRESS_FLAG_SUBNET16;
+                *(uint16_t*) ptr = gid->global.subnet_prefix >> 48;
+            }
         }
+    } else {
+        /* RoCE */
+        /* in this case we don't use the lid */
+        ib_addr->flags &= ~UCT_IB_ADDRESS_FLAG_LID;
+        ib_addr->flags |= UCT_IB_ADDRESS_FLAG_GID_RAW;
+        memcpy(ptr, gid->raw, 16 * sizeof(uint8_t));    /* uint8_t raw[16]; */
     }
+
 }
 
 void uct_ib_address_unpack(const uct_ib_address_t *ib_addr, uint16_t *lid,
                            uint8_t *is_global, union ibv_gid *gid)
 {
     const void *ptr = ib_addr + 1;
 
-    if (ib_addr->flags & UCT_IB_ADDRESS_FLAG_LID) {
-        *lid = *(uint16_t*)ptr;
-        ptr += sizeof(uint16_t);
-    } else {
+    if (ucs_unlikely(ib_addr->flags & UCT_IB_ADDRESS_FLAG_GID_RAW)) {
+        /* RoCE */
+        *is_global = 1;
         *lid = 0;
-    }
-
-    if (ib_addr->flags & UCT_IB_ADDRESS_FLAG_IF_ID) {
-        gid->global.interface_id = *(uint64_t*)ptr;
-        ptr += sizeof(uint64_t);
+        memcpy(gid->raw, ptr, 16 * sizeof(uint8_t));    /* uint8_t raw[16]; */
     } else {
-        gid->global.interface_id = 0;
-    }
+        /* IB */
```

Comment:
it is not present in the packed address unless the proper flag (which you check in the 'if') is set